src/services/codefixes/importFixes.ts

```diff
@@ -504,42 +504,60 @@ namespace ts.codefix {
-                        if (endsWith(relativeFileName, "/index")) {
-                            relativeFileName = getDirectoryPath(relativeFileName);
-                        }
-                        else {
-                            try {
-                                const moduleDirectory = getDirectoryPath(moduleFileName);
-                                const packageJsonContent = JSON.parse(context.host.readFile(combinePaths(moduleDirectory, "package.json")));
+                        // Simplify the full file path to something that can be resolved by Node.
+                        // First remove the extension
+                        let moduleSpecifier = removeFileExtension(moduleFileName);
+                        // If the module could be imported by a directory name, use that directory's name
+                        moduleSpecifier = getDirectoryOrFileName(moduleSpecifier);
+                        // Get a path that's relative to node_modules or the importing file's path
+                        moduleSpecifier = getNodeResolvablePath(moduleSpecifier);
+                        // If the module was found in @types, get the actual node package name
+                        return getPackageNameFromAtTypesDirectory(moduleSpecifier);
+
+                        function getDirectoryOrFileName(fullModulePathWithoutExtension: string): string {
+                            // If the file is the main module, it can be imported by the package name
+                            const indexOfLastNodeModules = moduleFileName.lastIndexOf("node_modules");
+                            const indexOfSlashAtPackageRoot = moduleFileName.indexOf("/", indexOfLastNodeModules + 13 /* "node_modules\".length */);
+                            const packageRootPath = moduleFileName.substring(0, indexOfSlashAtPackageRoot);
+                            const packageJsonPath = combinePaths(packageRootPath, "package.json");
+                            if (context.host.fileExists(packageJsonPath)) {
+                                const packageJsonContent = JSON.parse(context.host.readFile(packageJsonPath));
                                 if (packageJsonContent) {
-                                    const mainFile = packageJsonContent.main || packageJsonContent.typings;
-                                    if (mainFile) {
-                                        const mainExportFile = toPath(mainFile, moduleDirectory, getCanonicalFileName);
+                                    const mainFileRelative = packageJsonContent.typings || packageJsonContent.types || packageJsonContent.main;
+                                    if (mainFileRelative) {
+                                        const mainExportFile = toPath(mainFileRelative, packageRootPath, getCanonicalFileName);
                                         if (removeFileExtension(mainExportFile) === removeFileExtension(moduleFileName)) {

```

Comment:
Does this mean they're allowed to have different file extensions?  Also, does case matter?